netbsdsrc/usr.sbin/amd/fixmount/fixmount.8
.\"
.\" Copyright (c) 1997 Erez Zadok
.\" Copyright (c) 1990 Jan-Simon Pendry
.\" Copyright (c) 1990 Imperial College of Science, Technology & Medicine
.\" Copyright (c) 1990 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Jan-Simon Pendry at Imperial College, London.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgment:
.\"      This product includes software developed by the University of
.\"      California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	%W% (Berkeley) %G%
.\"
.\" $Id: fixmount.8,v 1.1.1.1 1997/12/22 22:28:42 cgd Exp $
.\"
.TH FIXMOUNT 8L "26 Feb 1993"
.SH NAME
fixmount \- fix remote mount entries
.SH SYNOPSIS
.B fixmount
[
.B \-adervq
]
[
.B \-h
.I name
]
.I host
\&...
.SH DESCRIPTION
.IX  "fixmount command"  ""  "\fLfixmount\fP \(em fix remote mount entries"
.LP
.B fixmount
is a variant of
.BR showmount (8)
that can delete bogus mount entries in remote
.BR mountd (8C)
daemons.
The actions specified by the options are performed for each
.I host
in turn.
.SH OPTIONS
.TP
.B \-a \-d \-e
These options work as in
.BR showmount (8)
except that only entries pertaining to the local host are printed.
.TP
.B \-r
Removes those remote mount entries on
.I host
that do not correspond to current mounts, i.e., which are left-over
from a crash or are the result of improper mount protocol.
The actuality of mounts is verified using the entries in
.BR /etc/mtab .
.TP
.B \-v
Verify remote mounts.  Similar to
.B \-r
except that only a notification message is printed for each bogus entry
found.  The remote mount table is not changed.
.TP
.B \-A
Issues a command to the remote mountd declaring that ALL of its filesystems 
have been unmounted.  This should be used with caution, as it removes all
remote mount entries pertaining to the local system, whether or not any
filesystems are still mounted locally.
.TP
.B \-q
Be quiet.
Suppresses error messages due to timeouts and "Program not registered",
i.e., due to remote hosts not supporting RPC or not running mountd.
.TP
.BI \-h \ name
Pretend the local hostname is
.IR name .
This is useful after the local hostname has been changed and rmtab entries
using the old name remain on a remote machine.
Unfortunately, most mountd's won't be able to successfully handle removal
of such entries, so this option is useful in combination with
.B \-v
only.
.br
This option also saves time as comparisons of remotely recorded and local
hostnames by address are avoided.
.SH FILES
.PD 0
.TP 20
.B /etc/mtab
List of current mounts.
.TP
.B /etc/rmtab
Backup file for remote mount entries on NFS server.
.PD
.SH "SEE ALSO"
.BR showmount (8),
.BR mtab (5),
.BR rmtab (5),
.BR mountd (8C).
.SH BUGS
No attempt is made to verify the information in
.B /etc/mtab
itself.
.PP
Since swap file mounts are not recorded in
.BR /etc/mtab ,
a heuristic specific to SunOS is used to determine whether such a mount
is actual (replacing the string "swap" with "root" and verifying the resulting
path).
.PP
Symbolic links on the server will cause the path in the remote entry to differ
from the one in
.BR /etc/mtab .
To catch those cases, a filesystem is also deemed mounted if its
.I local
mount point is identical to the remote entry.
I.e., on a SunOS diskless client,
.B server:/export/share/sunos.4.1.1
is actually
.BR /usr/share .
Since the local mount point is
.B /usr/share
as well this will be handled correctly.
.PP
There is no way to clear a stale entry in a remote mountd after the
local hostname (or whatever reverse name resolution returns for it)
has been changed.  To take care of these cases,
the remote /etc/rmtab file has to be edited and mountd restarted.
.PP
The RPC timeouts for mountd calls can only be changed by recompiling.
The defaults are 2 seconds for client handle creation and 5 seconds for
RPC calls.


ace/ChangeLog
Fri Oct 05 05:56:50 2001  Balachandran  <bala@cs.wustl.edu>

        * ACE version 5.2 released.

Thu Oct  4 17:38:21 2001  Steve Huston  <shuston@riverace.com>

        * ace/Synch.h: Added some comments to explain the rationale for
          the different return values from the blocking and timed
          versions of ACE_Null_Semaphore::acquire().  Thanks to Doug
          Schmidt for explaining this.

Thu Oct  4 17:27:39 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Handler.cpp (recv_log_record): Insure that
          the payload message block is properly aligned for any CDR
          extraction. Many thanks to Bala for noticing this problem and
          sending code for the fix.

Thu Oct  4 17:08:39 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Process_Per_Connection_Logging_Server.cpp:
          In prepare(), call ACE_Process_Options::command_line() rather
          than ACE_Process_Options::process_name() to set the program
          path for the child process.

Thu Oct  4 13:18:37 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Handler.cpp: Use a ACE_Message_Block,
          rather than hijacking a CDR block, to read the payload into.

Thu Oct  4 10:14:34 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* THANKS: Added Donald Acton to the Hall of Fame.

Thu Oct  4 06:39:20 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* ace/config-sunos5.5.h: Added ACE_LACKS_MKSTEMP as SunOS 5.5 does
	  not support mkstemp (). 
	
	* ace/config-sunos5.6.h: Undefined ACE_LACKS_MKSTEMP as Sun 5.6
	  and above support it. 

	  Thanks to Don Hinton <dhinton@swan.fr> for reporting this
	  problem and Chad Elliott for testing this out on a Sun OS 5.5
	  box. 

Wed Oct  3 16:37:41 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* ace/CDR_Stream.cpp (read_1): Reverted this change "Wed Oct  3
	  15:05:43 2001  Jeff Parsons <parsons@cs.wustl.edu>".  This
	  change is alone not enough to fix the problem that Brad
	  reported. The sample program for the text needs to be fixed to
	  get it working right.

Wed Oct  3 15:05:43 2001  Jeff Parsons <parsons@cs.wustl.edu>

        * ace/CDR_Stream.cpp (read_1);

          Reverted changed made in 

          Fri Aug 17 16:56:16 2001  Jeff Parsons <parsons@cs.wustl.edu>

          to read_1(). A similar change made in several other methods
          is ok, but in this case bytes can be copied into the
          input CDR stream without changing wr_ptr, so the check will
          fail unnecessarily. Thanks to Brad Hoskins <BHoskins@slo.newport.com>
          for pointing this out.

Wed Oct  2 12:18:37 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Server.cpp: Fixed spelling error.
        * examples/C++NPv1/Logging_Handler.cpp: Comment length adjustment.

Tue Oct  2 21:52:31 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Client.cpp: Reverted this:
          Mon Oct  1 18:27:41 2001  Steve Huston  <shuston@riverace.com>

Tue Oct  2 19:52:57 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* Makefile: Removed some of the files that no longer exist in the
	  repository. The inclusion of these files in the Makefile added
	  noise during the beta cutting process.

Tue Oct 02 15:34:49 2001  Carlos O'Ryan  <coryan@uci.edu>

	* etc/tao_cosevent.doxygen:
	* etc/tao_cosnotification.doxygen:
	  Fixed typos in the tag file names.

Tue Oct  2 15:52:27 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* THANKS: Added Sunny Leung to the Hall of Fame.

Tue Oct  2 15:33:36 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* THANKS: Removed one instance of Peter Phillips
	  <pphillip@opentext.com> from the Hall of Fame.

Tue Oct  2 15:00:37 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* ace/CDR_Stream.cpp (clone_from): Copied the GIOP major and minor
	  versions from the incoming CDR stream. Thanks to Sunny Leung 
	  <sleung@opentext.com> for reporting this. This closes bugzilla
	  #1043.  

Tue Oct  2 09:45:39 2001  Steve Huston  <shuston@riverace.com>

        * tests/libDLL_Test.icc:
        * tests/libService_Config_DLL.icc: Add link(exportAll) to the link
          options to export the entrypoints needed for the tests.

Mon Oct  1 20:04:14 2001  Priyanka Gontla  <pgontla@ece.uci.edu>

	* THANKS: 
	  Added Tushar Nair to the THANKS file.

Mon Oct  1 18:06:48 2001  Priyanka Gontla  <pgontla@ece.uci.edu>

	* ace/Object_Manager.cpp (ACE_UnhandledExceptionFilter):
	  Removed the un-needed #if 0's. Left them before as a safety
	  measure (in case something doesnt work) and forgot to remove
	  them later.

	* ace/Log_Msg.cpp (instance):
	  Fixed the seg faults that are raised when ace is compiled with
	  ACE_NTRACE=0 on SunOS. Added a check before locking and
	  releasing a lock.

	  Thanks to Tushar Nair <ntushar@yahoo.com> for reporting the
	  problem and for providing a fix.

Mon Oct  1 18:27:41 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Client.cpp: Remove the third (optional)
          arg from getline(). Using a char literal without further
          qualification causes confusion in Unicode builds.

Mon Oct  1 14:44:13 2001  Ossama Othman  <ossama@uci.edu>

	* ace/SSL/SSL_Context.cpp (ssl_library_init):

	  Initialize the locking callbacks in the crypto library before
	  initializing the OpenSSL library.  OpenSSL may require the
	  locking callbacks to be initialized before its own
	  initialization.

	* ace/SSL/SSL_Context.cpp (ssl_library_fini):

	  Perform crypto library resource cleanup before reclaiming the
	  locks used by the locking callbacks.  Crypto library resource
	  cleanup makes use of the locking callbacks.

Mon Oct  1 15:19:11 2001  Douglas C. Schmidt  <schmidt@macarena.cs.wustl.edu>


netbsdsrc/usr.sbin/amd/fixmount/fixmount.c
/*
 * Copyright (c) 1997 Erez Zadok
 * Copyright (c) 1990 Jan-Simon Pendry
 * Copyright (c) 1990 Imperial College of Science, Technology & Medicine
 * Copyright (c) 1990 The Regents of the University of California.
 * All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Jan-Simon Pendry at Imperial College, London.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by the University of
 *      California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *      %W% (Berkeley) %G%
 *
 * $Id: fixmount.c,v 1.1.1.1 1997/12/22 22:28:42 cgd Exp $
 *
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif /* HAVE_CONFIG_H */
#include <am_defs.h>

#define CREATE_TIMEOUT	2	/* seconds */
#define CALL_TIMEOUT	5	/* seconds */

#ifndef INADDR_NONE
# define INADDR_NONE	0xffffffff
#endif /* not INADDR_NONE */

/* Constant defs */
#define	ALL		1
#define	DIRS		2

#define	DODUMP		0x1
#define	DOEXPORTS	0x2
#define	DOREMOVE	0x4
#define	DOVERIFY	0x8
#define	DOREMALL	0x10

extern int fixmount_check_mount(char *host, struct in_addr hostaddr, char *path);

static char dir_path[NFS_MAXPATHLEN];
static char thishost[MAXHOSTNAMELEN] = "";
static exports mntexports;
static int quiet = 0;
static int type = 0;
static jmp_buf before_rpc;
static mountlist mntdump;
static struct in_addr thisaddr;
static CLIENT *clnt_create_timeout(char *, struct timeval *);

RETSIGTYPE create_timeout(int);
int is_same_host(char *, char *, struct in_addr);
int main(int, char *[]);
int remove_all(CLIENT *, char *);
int remove_mount(CLIENT *, char *, mountlist, int);
void fix_rmtab(CLIENT *, char *, mountlist, int, int);
void print_dump(mountlist);
void usage(void);

/* dummy variables */
char *progname;
char hostname[MAXHOSTNAMELEN];
int orig_umask, foreground, debug_flags;
pid_t mypid;
serv_state amd_state;

void
usage(void)
{
  fprintf(stderr, "usage: fixmount [-adervAqf] [-h hostname] host ...\n");
  exit(1);
}


/*
 * Check hostname against other name and its IP address
 */
int
is_same_host(char *name1, char *name2, struct in_addr addr2)
{
  if (strcasecmp(name1, name2) == 0) {
    return 1;
  } else if (addr2.s_addr == INADDR_NONE) {
    return 0;
  } else {
    static char lasthost[MAXHOSTNAMELEN] = "";
    static struct in_addr addr1;
    struct hostent *he;

    /*
     * To save nameserver lookups, and because this function
     * is typically called repeatedly on the same names,
     * cache the last lookup result and reuse it if possible.
     */
    if (strcasecmp(name1, lasthost) == 0) {
      return (addr1.s_addr == addr2.s_addr);
    } else if (!(he = gethostbyname(name1))) {
      return 0;
    } else {
      strncpy(lasthost, name1, sizeof(lasthost) - 1);
      memcpy(&addr1, he->h_addr, sizeof(addr1));
      return (addr1.s_addr == addr2.s_addr);
    }
  }
}


/*
 * Print the binary tree in inorder so that output is sorted.
 */
void
print_dump(mountlist mp)
{
  if (mp == NULL)
    return;
  if (is_same_host(mp->ml_hostname, thishost, thisaddr)) {
    switch (type) {
    case ALL:
      printf("%s:%s\n", mp->ml_hostname, mp->ml_directory);
      break;
    case DIRS:
      printf("%s\n", mp->ml_directory);
      break;
    default:
      printf("%s\n", mp->ml_hostname);
      break;
    };
  }
  if (mp->ml_next)
    print_dump(mp->ml_next);
}


/*
 * remove entry from remote rmtab
 */
int
remove_mount(CLIENT *client, char *host, mountlist ml, int fixit)
{
  enum clnt_stat estat;
  struct timeval tv;
  char *pathp = dir_path;

  strncpy(dir_path, ml->ml_directory, sizeof(dir_path));

  if (!fixit) {
    printf("%s: bogus mount %s:%s\n", host, ml->ml_hostname, ml->ml_directory);
    fflush(stdout);
  } else {
    printf("%s: removing %s:%s\n", host, ml->ml_hostname, ml->ml_directory);
    fflush(stdout);

    tv.tv_sec = CALL_TIMEOUT;
    tv.tv_usec = 0;

    if ((estat = clnt_call(client,
			   MOUNTPROC_UMNT,
			   (XDRPROC_T_TYPE) xdr_dirpath,
			   (char *) &pathp,
			   (XDRPROC_T_TYPE) xdr_void,
			   (char *) 0,
			   tv)) != RPC_SUCCESS) {
      fprintf(stderr, "%s:%s MOUNTPROC_UMNT: ",
	      host, ml->ml_directory);
      clnt_perrno(estat);
      fflush(stderr);
      return -1;
    }
  }
  return 0;
}


/*
 * fix mount list on remote host
 */
void
fix_rmtab(CLIENT *client, char *host, mountlist mp, int fixit, int force)
{
  mountlist p;
  struct hostent *he;
  struct in_addr hostaddr;

  /*
   * Obtain remote address for comparisons
   */
  if ((he = gethostbyname(host))) {
    memcpy(&hostaddr, he->h_addr, sizeof(hostaddr));
  } else {
    hostaddr.s_addr = INADDR_NONE;
  }

  for (p = mp; p; p = p->ml_next) {
    if (is_same_host(p->ml_hostname, thishost, thisaddr)) {
      if (force || !fixmount_check_mount(host, hostaddr, p->ml_directory))
	remove_mount(client, host, p, fixit);
    }
  }
}


/*
 * remove all entries from remote rmtab
 */
int
remove_all(CLIENT *client, char *host)
{
  enum clnt_stat estat;
  struct timeval tv;

  printf("%s: removing ALL\n", host);
  fflush(stdout);

  tv.tv_sec = CALL_TIMEOUT;
  tv.tv_usec = 0;

  if ((estat = clnt_call(client,
			 MOUNTPROC_UMNTALL,
			 (XDRPROC_T_TYPE) xdr_void,
			 (char *) 0,
			 (XDRPROC_T_TYPE) xdr_void,
			 (char *) 0,
			 tv)) != RPC_SUCCESS) {
    /*
     * RPC_SYSTEMERROR is returned even if all went well
     */
    if (estat != RPC_SYSTEMERROR) {
      fprintf(stderr, "%s MOUNTPROC_UMNTALL: ", host);
      clnt_perrno(estat);
      fflush(stderr);
      return -1;
    }
  }

  return 0;
}


/*
 * This command queries the NFS mount daemon for it's mount list and/or
 * it's exports list and prints them out.
 * See "NFS: Network File System Protocol Specification, RFC1094, Appendix A"
 * for detailed information on the protocol.
 */
int
main(int argc, char *argv[])
{
  AUTH *auth;
  CLIENT *client;
  char *host;
  enum clnt_stat estat;
  exports exp;
  extern char *optarg;
  extern int optind;
  groups grp;
  int ch;
  int force = 0;
  int morethanone;
  register int rpcs = 0;
  struct timeval tv;

  while ((ch = getopt(argc, argv, "adervAqfh:")) != -1)
    switch ((char) ch) {

    case 'a':
      if (type == 0) {
	type = ALL;
	rpcs |= DODUMP;
      } else
	usage();
      break;

    case 'd':
      if (type == 0) {
	type = DIRS;
	rpcs |= DODUMP;
      } else
	usage();
      break;

    case 'e':
      rpcs |= DOEXPORTS;
      break;

    case 'r':
      rpcs |= DOREMOVE;
      break;

    case 'A':
      rpcs |= DOREMALL;
      break;

    case 'v':
      rpcs |= DOVERIFY;
      break;

    case 'q':
      quiet = 1;
      break;

    case 'f':
      force = 1;
      break;

    case 'h':
      strncpy(thishost, optarg, sizeof(thishost));
      thishost[sizeof(thishost) - 1] = '\0';
      break;

    case '?':
    default:
      usage();
    }

  if (optind == argc)
    usage();

  if (rpcs == 0)
    rpcs = DODUMP;

  if (!*thishost) {
    struct hostent *he;

    if (gethostname(thishost, sizeof(thishost)) < 0) {
      perror("gethostname");
      exit(1);
    }

    /*
     * We need the hostname as it appears to the other side's
     * mountd, so get our own hostname by reverse address
     * resolution.
     */
    if (!(he = gethostbyname(thishost))) {
      fprintf(stderr, "gethostbyname failed on %s\n",
	      thishost);
      exit(1);
    }
    memcpy(&thisaddr, he->h_addr, sizeof(thisaddr));
    if (!(he = gethostbyaddr((char *) &thisaddr, sizeof(thisaddr),
			     he->h_addrtype))) {
      fprintf(stderr, "gethostbyaddr failed on %s\n",
	      inet_ntoa(thisaddr));
      exit(1);
    }
    strncpy(thishost, he->h_name, sizeof(thishost));
    thishost[sizeof(thishost) - 1] = '\0';
  } else {
    thisaddr.s_addr = INADDR_NONE;
  }

  if (!(auth = authunix_create_default())) {
    fprintf(stderr, "couldn't create authentication handle\n");
    exit(1);
  }
  morethanone = (optind + 1 < argc);

  for (; optind < argc; optind++) {

    host = argv[optind];
    tv.tv_sec = CREATE_TIMEOUT;
    tv.tv_usec = 0;

    if (!(client = clnt_create_timeout(host, &tv)))
        continue;

    client->cl_auth = auth;
    tv.tv_sec = CALL_TIMEOUT;
    tv.tv_usec = 0;

    if (rpcs & (DODUMP | DOREMOVE | DOVERIFY))
      if ((estat = clnt_call(client,
			     MOUNTPROC_DUMP,
			     (XDRPROC_T_TYPE) xdr_void,
			     (char *) 0,
			     (XDRPROC_T_TYPE) xdr_mountlist,
			     (char *) &mntdump,
			     tv)) != RPC_SUCCESS) {
	fprintf(stderr, "%s: MOUNTPROC_DUMP: ", host);
	clnt_perrno(estat);
	fflush(stderr);
	mntdump = NULL;
	goto next;
      }
    if (rpcs & DOEXPORTS)
      if ((estat = clnt_call(client,
			     MOUNTPROC_EXPORT,
			     (XDRPROC_T_TYPE) xdr_void,
			     (char *) 0,
			     (XDRPROC_T_TYPE) xdr_exports,
			     (char *) &mntexports,
			     tv)) != RPC_SUCCESS) {
	fprintf(stderr, "%s: MOUNTPROC_EXPORT: ", host);
	clnt_perrno(estat);
	fflush(stderr);
	mntexports = NULL;
	goto next;
      }

    /* Now just print out the results */
    if ((rpcs & (DODUMP | DOEXPORTS)) &&
	morethanone) {
      printf(">>> %s <<<\n", host);
      fflush(stdout);
    }

    if (rpcs & DODUMP) {
      print_dump(mntdump);
    }

    if (rpcs & DOEXPORTS) {
      exp = mntexports;
      while (exp) {
	printf("%-35s", exp->ex_dir);
	grp = exp->ex_groups;
	if (grp == NULL) {
	  printf("Everyone\n");
	} else {
	  while (grp) {
	    printf("%s ", grp->gr_name);
	    grp = grp->gr_next;
	  }
	  printf("\n");
	}
	exp = exp->ex_next;
      }
    }

    if (rpcs & DOVERIFY)
      fix_rmtab(client, host, mntdump, 0, force);

    if (rpcs & DOREMOVE)
      fix_rmtab(client, host, mntdump, 1, force);

    if (rpcs & DOREMALL)
      remove_all(client, host);

  next:
    if (mntdump)
      (void) clnt_freeres(client,
			  (XDRPROC_T_TYPE) xdr_mountlist,
			  (char *) &mntdump);
    if (mntexports)
      (void) clnt_freeres(client,
			  (XDRPROC_T_TYPE) xdr_exports,
			  (char *) &mntexports);

    clnt_destroy(client);
  }
  exit(0);
  return 0; /* should never reach here */
}


RETSIGTYPE
create_timeout(int sig)
{
  signal(SIGALRM, SIG_DFL);
  longjmp(before_rpc, 1);
}


static CLIENT *
clnt_create_timeout(char *host, struct timeval *to)
{
  CLIENT *client;

  if (setjmp(before_rpc)) {
    if (!quiet) {
      fprintf(stderr, "%s: ", host);
      clnt_perrno(RPC_TIMEDOUT);
      fprintf(stderr, "\n");
      fflush(stderr);
    }
    return NULL;
  }
  signal(SIGALRM, create_timeout);
  ualarm(to->tv_sec * 1000000 + to->tv_usec, 0);

  /*
   * Try TCP first (in case return data is large), then UDP
   */
  if (!(client = clnt_create(host, MOUNTPROG, MOUNTVERS, "tcp")) &&
      ! (client = clnt_create(host, MOUNTPROG, MOUNTVERS, "udp"))) {
    ualarm(0, 0);
    if (!quiet) {
      clnt_pcreateerror(host);
      fflush(stderr);
    }
    return NULL;
  }
  ualarm(0, 0);
  return client;
}
